fundamentals/android/auth-anonymous/Notes/app/src/main/java/com/notes/app/model/service/impl/AccountServiceImpl.kt

```diff
@@ -30,16 +31,31 @@ class AccountServiceImpl @Inject constructor() : AccountService {
         return Firebase.auth.currentUser != null
     }
 
-    override suspend fun signIn(email: String, password: String) {
-        Firebase.auth.signInWithEmailAndPassword(email, password).await()
+    override fun isAnonymousUser(): Boolean {
+        return Firebase.auth.currentUser?.isAnonymous ?: true
+    }
+
+    override suspend fun createAnonymousAccount() {
+        Firebase.auth.signInAnonymously().await()
+    }
+
+    override suspend fun linkAccount(email: String, password: String) {
+        val credential = EmailAuthProvider.getCredential(email, password)
+        Firebase.auth.currentUser!!.linkWithCredential(credential).await()
     }
 
-    override suspend fun signUp(email: String, password: String) {
-        Firebase.auth.createUserWithEmailAndPassword(email, password).await()
+    override suspend fun signIn(email: String, password: String) {
+        Firebase.auth.signInWithEmailAndPassword(email, password).await()
     }
 
     override suspend fun signOut() {
+        if (Firebase.auth.currentUser!!.isAnonymous) {
+            Firebase.auth.currentUser!!.delete()
```

Comment:
I'm curious: Why do we delete anonymous accounts?

Reply:
Because you can never sign back in to an anonymous account. But it just occurred to me that this code is unreachable, as the sign out option is only shown to non-anonymous accounts. Removed this piece of code.

Reply:
Can't you unlink providers (https://firebase.google.com/docs/auth/ios/account-linking#unlink-an-auth-provider-from-a-user-account), and thus get back into a state where the account essentially is anonymous?